File: sdk/typescript/tests/identity.test.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 */

import { describe, expect, it } from "vitest";

import {
  InvalidArgumentsError,
  InvalidTokenError,
  Plinth,
  SigningKeyNotFoundError,
  TokenExpiredError,
  TokenRevokedError,
  UnauthorizedError,
  type SigningKey,
  type TokenClaims,
  type TokenIssueResponse,
} from "../src/index.js";
import { MockServer } from "./_helpers.js";

function makeClient(server: MockServer, identityUrl = "http://identity.test"): Plinth {
  return new Plinth({
    workspaceUrl: "http://workspace.test",
    gatewayUrl: "http://gateway.test",
    identityUrl,
    apiKey: "bootstrap-token",
    fetch: server.fetch as unknown as typeof fetch,
  });
}

const sampleClaims: TokenClaims = {
  sub: "agent_1",
  iss: "http://identity.test",
  aud: "plinth",
  iat: 1_700_000_000,
  exp: 1_700_003_600,
  jti: "jti_abc",
  agent_id: "agent_1",
  tenant_id: "default",
  workspace_id: null,
  scopes: ["tool:web.fetch:read"],
  rate_limit: null,
};

describe("IdentityClient — issue / verify / revoke", () => {
  it("issueToken POSTs the request body and returns the response", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      expect(body.agent_id).toBe("agent_1");
      expect(body.scopes).toEqual(["tool:web.fetch:read", "workspace:my-task:write"]);
      expect(body.tenant_id).toBe("default");
      expect(body.ttl_seconds).toBe(3600);
      return {
        status: 201,
        body: {
          token: "header.payload.sig",
          jti: "jti_abc",
          expires_at: "2026-01-01T01:00:00Z",
          claims: sampleClaims,
        } satisfies TokenIssueResponse,
      };
    });

    const client = makeClient(server);
    const issued = await client.identity.issueToken({
      agentId: "agent_1",
      scopes: ["tool:web.fetch:read", "workspace:my-task:write"],
      tenantId: "default",
      ttlSeconds: 3600,
    });
    expect(issued.token).toBe("header.payload.sig");
    expect(issued.claims.agent_id).toBe("agent_1");
  });

  it("issueToken omits optional fields not provided", async () => {
    const server = new MockServer();
    let captured: Record<string, unknown> = {};
    server.on("POST", /\/v1\/tokens$/, (req) => {
      captured = JSON.parse(req.body ?? "{}");
      return {
        status: 201,
        body: {
          token: "t",
          jti: "j",
          expires_at: "2026-01-01T00:00:00Z",
          claims: sampleClaims,
        },
      };
    });
    const client = makeClient(server);
    await client.identity.issueToken({ agentId: "a", scopes: ["s"] });
    expect(captured).toEqual({ agent_id: "a", scopes: ["s"] });
  });

  it("verifyToken POSTs the token and returns the claims", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens\/verify$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      expect(body.token).toBe("xyz");
      return { body: sampleClaims };
    });
    const client = makeClient(server);
    const claims = await client.identity.verifyToken("xyz");
    expect(claims.jti).toBe("jti_abc");
    expect(claims.scopes).toEqual(["tool:web.fetch:read"]);
  });

  it("revokeToken POSTs to the revoke endpoint", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens\/jti_abc\/revoke$/, () => ({ status: 204 }));
    const client = makeClient(server);
    await expect(client.identity.revokeToken("jti_abc")).resolves.toBeUndefined();
  });

  it("getTokenInfo returns the token metadata view", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/tokens\/jti_abc$/, {
      jti: "jti_abc",
      agent_id: "agent_1",
      tenant_id: "default",
      issued_at: "2026-01-01T00:00:00Z",
      expires_at: "2026-01-01T01:00:00Z",
      revoked: false,
      revoked_at: null,
      metadata: {},
    });
    const client = makeClient(server);
    const info = await client.identity.getTokenInfo("jti_abc");
    expect(info.revoked).toBe(false);
    expect(info.agent_id).toBe("agent_1");
  });

  it("getJwks returns the well-known JSON document", async () => {
    const server = new MockServer();
    server.json("GET", /\/\.well-known\/jwks\.json$/, { keys: [{ kid: "k1", alg: "RS256" }] });
    const client = makeClient(server);
    const jwks = await client.identity.getJwks();
    expect(Array.isArray((jwks as { keys: unknown[] }).keys)).toBe(true);
  });
});

describe("IdentityClient — error mapping", () => {
  it("maps INVALID_TOKEN to InvalidTokenError (subclass of UnauthorizedError)", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens\/verify$/, () => ({
      status: 401,
      body: { error: { code: "INVALID_TOKEN", message: "bad token" } },
    }));
    const client = makeClient(server);
    const err = await client.identity
      .verifyToken("bogus")
      .catch((e: unknown) => e as Error);
    expect(err).toBeInstanceOf(InvalidTokenError);
    expect(err).toBeInstanceOf(UnauthorizedError);
    expect((err as InvalidTokenError).code).toBe("INVALID_TOKEN");
  });

  it("maps TOKEN_EXPIRED to TokenExpiredError", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens\/verify$/, () => ({
      status: 401,
      body: { error: { code: "TOKEN_EXPIRED", message: "expired" } },
    }));
    const client = makeClient(server);
    await expect(client.identity.verifyToken("x")).rejects.toBeInstanceOf(TokenExpiredError);
  });

  it("maps TOKEN_REVOKED to TokenRevokedError", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/tokens\/verify$/, () => ({
      status: 401,
      body: { error: { code: "TOKEN_REVOKED", message: "revoked" } },
    }));
    const client = makeClient(server);
    await expect(client.identity.verifyToken("x")).rejects.toBeInstanceOf(TokenRevokedError);
  });
});

describe("Plinth.identity — accessor", () => {
  it("throws when identityUrl was not configured", () => {
    const server = new MockServer();
    const client = new Plinth({
      workspaceUrl: "http://workspace.test",
      gatewayUrl: "http://gateway.test",
      apiKey: "bootstrap",
      fetch: server.fetch as unknown as typeof fetch,
    });
    expect(() => client.identity).toThrow(/identityUrl/);
  });
});

// ---------------------------------------------------------------------------
// v0.4 — Signing keys

const sampleKey: SigningKey = {
  kid: "abc1234567890def",
  alg: "RS256",
  public_key_pem: "-----BEGIN PUBLIC KEY-----\nfake\n-----END PUBLIC KEY-----\n",
  created_at: "2026-01-01T00:00:00Z",
  rotated_in_at: "2026-01-01T00:00:00Z",
  expires_at: "2026-03-01T00:00:00Z",
  active: true,
};

describe("IdentityClient — signing keys (v0.4)", () => {
  it("listKeys returns the keys array", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/keys$/, { keys: [sampleKey] });
    const client = makeClient(server);
    const keys = await client.identity.listKeys();
    expect(keys).toHaveLength(1);
    expect(keys[0].kid).toBe("abc1234567890def");
    expect(keys[0].active).toBe(true);
  });

  it("listKeys passes include_expired when requested", async () => {
    const server = new MockServer();
    let capturedUrl = "";
    server.on("GET", /\/v1\/keys/, (req) => {
      capturedUrl = req.url;
      return { body: { keys: [] } };
    });
    const client = makeClient(server);
    await client.identity.listKeys({ includeExpired: true });
    expect(capturedUrl).toContain("include_expired=true");
  });

  it("listKeys returns empty list for HS256 deployments", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/keys/, { keys: [] });
    const client = makeClient(server);
    const keys = await client.identity.listKeys();
    expect(keys).toEqual([]);
  });

  it("rotateKey returns the new active key", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/keys\/rotate$/, () => ({
      status: 201,
      body: { ...sampleKey, kid: "newkid000000abcd" },
    }));
    const client = makeClient(server);
    const key = await client.identity.rotateKey();
    expect(key.kid).toBe("newkid000000abcd");
    expect(key.active).toBe(true);
  });

  it("rotateKey throws InvalidArgumentsError on HS256 deployments", async () => {
    const server = new MockServer();
    server.on("POST", /\/v1\/keys\/rotate$/, () => ({
      status: 400,
      body: {
        error: {
          code: "INVALID_ARGUMENTS",
          message: "key rotation is only available when jwt_alg=RS256",
          details: { jwt_alg: "HS256" },
        },
      },
    }));
    const client = makeClient(server);
    await expect(client.identity.rotateKey()).rejects.toBeInstanceOf(InvalidArgumentsError);
  });

  it("expireKey DELETEs the key", async () => {
    const server = new MockServer();
    let captured = "";
    server.on("DELETE", /\/v1\/keys\/abc1234567890def$/, (req) => {
      captured = req.method;
      return { status: 204 };
    });
    const client = makeClient(server);
    await expect(client.identity.expireKey("abc1234567890def")).resolves.toBeUndefined();
    expect(captured).toBe("DELETE");
  });

  it("expireKey unknown kid → SigningKeyNotFoundError", async () => {
    const server = new MockServer();
    server.on("DELETE", /\/v1\/keys\/.+$/, () => ({
      status: 404,
      body: {
        error: {
          code: "SIGNING_KEY_NOT_FOUND",
          message: "Signing key 'nope' does not exist",
          details: { kid: "nope" },
        },
      },
    }));
    const client = makeClient(server);
    await expect(client.identity.expireKey("nope")).rejects.toBeInstanceOf(
      SigningKeyNotFoundError,
    );
  });

  it("getKey returns matching key from the list", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/keys/, {
      keys: [
        sampleKey,
        { ...sampleKey, kid: "second", active: false },
      ],
    });
    const client = makeClient(server);
    const key = await client.identity.getKey("second");
    expect(key.kid).toBe("second");
    expect(key.active).toBe(false);
  });

  it("getKey throws SigningKeyNotFoundError when the kid is missing", async () => {
    const server = new MockServer();
    server.json("GET", /\/v1\/keys/, { keys: [] });
    const client = makeClient(server);
    await expect(client.identity.getKey("missing")).rejects.toBeInstanceOf(
      SigningKeyNotFoundError,
    );
  });
});

File: sdk/go/plinth_test/identity_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth_test

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/plinth/sdk-go/plinth"
)

// TestIdentityIssueToken covers the happy path: a valid request flows
// through, the JWT + decoded claims come back, and the JTI is exposed.
func TestIdentityIssueToken(t *testing.T) {
	ms := NewMockServer(t)
	expires := time.Now().Add(1 * time.Hour).Format(time.RFC3339Nano)
	ms.JSON("POST", "/v1/tokens", 201, map[string]any{
		"token":      "eyJ.fake.jwt",
		"jti":        "tok_1",
		"expires_at": expires,
		"claims": map[string]any{
			"sub":       "agent-A",
			"iss":       "http://localhost:7425",
			"aud":       "plinth",
			"iat":       1,
			"exp":       2,
			"jti":       "tok_1",
			"agent_id":  "agent-A",
			"tenant_id": "default",
			"scopes":    []any{"tool:web.fetch:read"},
		},
	})

	c := newTestClient(t, ms)
	id, err := c.IdentityClient()
	if err != nil {
		t.Fatalf("IdentityClient: %v", err)
	}
	resp, err := id.IssueToken(context.Background(), plinth.TokenIssueRequest{
		AgentID:    "agent-A",
		Scopes:     []string{"tool:web.fetch:read"},
		TTLSeconds: 3600,
	})
	if err != nil {
		t.Fatalf("IssueToken: %v", err)
	}
	if resp.JTI != "tok_1" {
		t.Errorf("JTI = %q, want tok_1", resp.JTI)
	}
	if resp.Token != "eyJ.fake.jwt" {
		t.Errorf("Token = %q, want eyJ.fake.jwt", resp.Token)
	}
	if len(resp.Claims.Scopes) != 1 || resp.Claims.Scopes[0] != "tool:web.fetch:read" {
		t.Errorf("Claims.Scopes = %v, want [tool:web.fetch:read]", resp.Claims.Scopes)
	}
}

// TestIdentityVerifyToken covers the verify path.
func TestIdentityVerifyToken(t *testing.T) {
	ms := NewMockServer(t)
	ms.JSON("POST", "/v1/tokens/verify", 200, map[string]any{
		"sub":       "agent-A",
		"iss":       "http://localhost:7425",
		"aud":       "plinth",
		"iat":       1,
		"exp":       2,
		"jti":       "tok_1",
		"agent_id":  "agent-A",
		"tenant_id": "default",
		"scopes":    []any{"workspace:ws_x:read"},
	})

	c := newTestClient(t, ms)
	id, _ := c.IdentityClient()
	claims, err := id.VerifyToken(context.Background(), "eyJ.fake.jwt")
	if err != nil {
		t.Fatalf("VerifyToken: %v", err)
	}
	if claims.JTI != "tok_1" {
		t.Errorf("JTI = %q, want tok_1", claims.JTI)
	}
}

// TestIdentityVerifyExpiredSurfacesTyped covers the typed-error path:
// a TOKEN_EXPIRED envelope should match ErrTokenExpired.
func TestIdentityVerifyExpiredSurfacesTyped(t *testing.T) {
	ms := NewMockServer(t)
	ms.Error("POST", "/v1/tokens/verify", 401, "TOKEN_EXPIRED", "exp")
	c := newTestClient(t, ms)
	id, _ := c.IdentityClient()
	_, err := id.VerifyToken(context.Background(), "expired")
	if !errors.Is(err, plinth.ErrTokenExpired) {
		t.Errorf("err = %v, want ErrTokenExpired", err)
	}
	// Also satisfies the broader Unauthorized matcher? In our model
	// ErrTokenExpired and ErrUnauthorized are independent sentinels —
	// callers wanting to catch "any auth failure" should branch on
	// status code 401 or use multiple errors.Is checks. Not asserted
	// here.
}

// TestIdentityRevokeToken covers the DELETE-style revoke endpoint.
func TestIdentityRevokeToken(t *testing.T) {
	ms := NewMockServer(t)
	ms.On("DELETE", "/v1/tokens/tok_1/revoke", func(_ recordedRequest) mockResponse {
		return mockResponse{Status: 204}
	})
	c := newTestClient(t, ms)
	id, _ := c.IdentityClient()
	if err := id.RevokeToken(context.Background(), "tok_1"); err != nil {
		t.Fatalf("RevokeToken: %v", err)
	}
}

// TestIdentityListKeys covers the v0.4 signing-keys list.
func TestIdentityListKeys(t *testing.T) {
	ms := NewMockServer(t)
	now := time.Now().Format(time.RFC3339Nano)
	exp := time.Now().Add(30 * 24 * time.Hour).Format(time.RFC3339Nano)
	ms.JSON("GET", "/v1/keys", 200, map[string]any{
		"keys": []any{
			map[string]any{
				"kid":            "k1",
				"alg":            "RS256",
				"public_key_pem": "-----BEGIN…",
				"created_at":     now,
				"expires_at":     exp,
				"active":         true,
			},
		},
	})
	c := newTestClient(t, ms)
	id, _ := c.IdentityClient()
	keys, err := id.ListKeys(context.Background(), false)
	if err != nil {
		t.Fatalf("ListKeys: %v", err)
	}
	if len(keys) != 1 || keys[0].KID != "k1" || !keys[0].Active {
		t.Errorf("keys = %+v, want one active k1", keys)
	}
}

// TestIdentityNotConfiguredOnPlinth verifies that IdentityClient()
// fails fast when Config.IdentityURL was empty.
func TestIdentityNotConfiguredOnPlinth(t *testing.T) {
	c, err := plinth.New(plinth.Config{
		WorkspaceURL: "http://localhost:7421",
		GatewayURL:   "http://localhost:7422",
		APIKey:       "x",
	})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	if _, err := c.IdentityClient(); !errors.Is(err, plinth.ErrIdentityNotConfigured) {
		t.Errorf("err = %v, want ErrIdentityNotConfigured", err)
	}
}
